### PR TITLE
Consolidate variant implementations

### PR DESCRIPTION
**Pull request by mavam**
This PR converts the various sum types in VAST from `vast::variant` or `std::variant` to `caf::variant`.

Types to convert:
- [x] `vast::json`
- [x] `vast::view`
- [x] `vast::bitmap`
- [x] `vast::expression`
- [ ] `vast::data`
- [ ] `vast::type`

Cleanup:
- [x] Make sure all variant types model the `SumType` concept in CAF.
- [ ] Remove `vast::variant`

### CODE REVIEW

**Comment by mavam** [comment]
This PR depends on actor-framework/actor-framework#697.

**Comment by mavam** [comment]
I have some issues with the current state. The new `json` class has just a variant as member. Technically, this variant *is* the API, because a JSON value is a sum type. However, we cannot use simply
```cpp
using json = caf::variant<Ts...>;
```
because this does't work with recursion. For example, the array type needs to take a sequence of JSON objects, i.e., a `vector<json>`. By wrapping the variant inside the `json` class, we enable recursion. Actually, for other sum types, which have additional attributes like `name()`, this design is a natural fit. But for the case of JSON, the sum type *is* the API. There's nothing else there but the data.

One option to model this is making the member simply public. Then `json x; x.value` would provide the variant access. Another option is to make `json` inherit from the variant. This is ugly because (1) inheritance itself, but more so (2) we cannot use the variant free functions anymore, e.g., `get` and `holds_alternative`.

That's why the current rewrite simply exposes a `.value()` member that allows the user work with the variant. If anyone has better ideas, please let me know.

File: libvast/vast/bitmap.hpp [hunk 1]
```diff
@@ -13,26 +13,35 @@
...
+
+public:
+  using types = caf::detail::type_list<
     ewah_bitmap,
     null_bitmap,
     wah_bitmap
   >;
 
-public:
+  using variant = caf::detail::tl_apply_t<types, caf::variant>;
```
Comment (Neverlord) [inline libvast/vast/bitmap.hpp]: Nitpick: I'd use `variant_type` here instead.

**Comment by mavam** [comment]
I've restored the branch because this PR was not yet done. I'm still working on porting `type` and `data`, the two remaining pieces. The former will no longer be a variant and the double-dispatching needs to be adapted in line with #199.

**Pull request merged**